src/core/expression/qgsexpressionfunction.cpp

```diff
@@ -262,6 +263,30 @@ bool QgsExpressionFunction::allParamsStatic( const QgsExpressionNodeFunction *no
   return true;
 }
 
+/**
+ *
+ * @param layerNode
+ * @param parent
+ * @param context
+ * @return
+ */
+bool prepareLayerNode( QgsExpressionNode *layerNode, QgsExpression *parent, const QgsExpressionContext *context )
+{
+  if ( layerNode->isStatic( parent, context ) )
+  {
+    QVariant layer = layerNode->eval( parent, context );
+    bool foundLayer = false;
+
+    std::shared_ptr<QgsVectorLayerFeatureSource> featureSource = QgsExpressionUtils::getFeatureSource( layer, context, parent, foundLayer );
```

Comment:
Hmm - i have my doubts that it's safe to share QgsVectorLayerFeatureSources without some protection. We can safely create a source in one thread and then hand it off to another thread, but we can't safely use the same source from multiple threads simultaneously (which I gather this is doing?).

Reply:
How it is used in the current approach:

1. on one thread a source is prepared (ideally on the main thread)
2. the source is used from one thread only

Do you think this needs more protection and if yes, in what way?

How I understand things: prepare is allowed to modify the expression context and expression nodes. These are in turn used by all evaluations of the expression.
We don't have any protective measures in place to limit the usage of a prepared expression to a specific thread. I also don't think we reviewed all code to make sure the access patterns to expression context/nodes during expression evaluation are thread safe. Am I missing something or is this actually a broader discussion about multithreading concepts for expressions?

Reply:
> 2. the source is used from one thread only
> Do you think this needs more protection and if yes, in what way?

There's no guarantees here that the source will be used from one thread. I think if we had some mutex or similar protecting access to to the shared source pointer then we should be ok.

Reply:
This would need to be on the getFeatures call, correct?

Reply:
Yeah, after we get an iterator we can be confident we're being thread safe